mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.james.mailbox.cassandra.CassandraMessageId;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.exception.MailboxNotFoundException;
+import org.apache.james.mailbox.model.ComposedMessageId;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.store.mail.AttachmentMapper;
+import org.apache.james.mailbox.store.mail.MailboxMapper;
+import org.apache.james.mailbox.store.mail.MessageIdMapper;
+import org.apache.james.mailbox.store.mail.MessageMapper.FetchType;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+import org.apache.james.mailbox.store.mail.model.Message;
+
+import com.datastax.driver.core.Row;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Throwables;
+
+public class CassandraMessageIdMapper implements MessageIdMapper {
+
+    private final AttachmentMapper attachmentMapper;
+    private final MailboxMapper mailboxMapper;
+    private final CassandraImapUidDAO imapUidDAO;
+    private final CassandraMessageIdDAO messageIdDAO;
+    private final CassandraMessageDAO messageDAO;
+
+    public CassandraMessageIdMapper(AttachmentMapper attachmentMapper, MailboxMapper mailboxMapper,
+            CassandraImapUidDAO imapUidDAO, CassandraMessageIdDAO messageIdDAO, CassandraMessageDAO messageDAO) {
+        this.attachmentMapper = attachmentMapper;
+        this.mailboxMapper = mailboxMapper;
+        this.imapUidDAO = imapUidDAO;
+        this.messageIdDAO = messageIdDAO;
+        this.messageDAO = messageDAO;
+    }
+
+    @Override
+    public List<Message> find(List<MessageId> messageIds, FetchType fetchType) {
+        List<CassandraMessageId> cassandraMessageIds = messageIds.stream()
+            .map(id -> (CassandraMessageId) id)
+            .collect(Collectors.toList());
+        return CassandraUtils.convertToStream(messageDAO.retrieveMessages(cassandraMessageIds, fetchType, Optional.empty()).join())
+                .map(row -> message(row, fetchType))
+                .sorted(Comparator.comparing(MailboxMessage::getUid))
+                .collect(Guavate.toImmutableList());
+    }
+
+    private MailboxMessage message(Row row, FetchType fetchType) {
+        try {
+            return CassandraMessageRowHandler.builder()
+                        .row(row)
+                        .messageId(retrieveUniqueMessageId(CassandraMessageId.of(row.getUUID(MESSAGE_ID))))
+                        .loadingAttachmentsFunction(attachmentMapper::getAttachments)
+                        .build()
+                    .toMailboxMessage(fetchType);
+        } catch (MailboxException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private ComposedMessageId retrieveUniqueMessageId(CassandraMessageId messageId) throws MailboxException {
+        return imapUidDAO.retrieve(messageId, Optional.empty())
+            .findFirst()
```

Comment:
I'm not comfortable with this find first, I fear of consequances if MessageId is duplicated. What about returning a list + flatMap ?